cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,183 @@
+  __host__ __device__ minmax_pair<T> operator()(const minmax_pair<T> &x,
+                                                const minmax_pair<T> &y) const
+  {
+    T x_min = x.min_valid ? x.min_val : cudf::DeviceMin::identity<T>();
+    T y_min = y.min_valid ? y.min_val : cudf::DeviceMin::identity<T>();
+    T x_max = x.max_valid ? x.max_val : cudf::DeviceMax::identity<T>();
+    T y_max = y.max_valid ? y.max_val : cudf::DeviceMax::identity<T>();
+
+    // The only invalid situation is if we compare two invalid values.
+    // Otherwise, we are certain to select a valid value due to the
+    // identity functions above changing the comparison value.
+    bool valid_min_result = x.min_valid || y.min_valid;
+    bool valid_max_result = x.max_valid || y.max_valid;
+
+    return minmax_pair<T>{
+      thrust::min(x_min, y_min), valid_min_result, thrust::max(x_max, y_max), valid_max_result};
+  }
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Expects
+ * no null values.
+ *
+ */
+template <typename T>
+struct minmax_no_null_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __host__ __device__ minmax_pair<T> operator()(const minmax_pair<T> &x,
+                                                const minmax_pair<T> &y) const
+  {
+    return minmax_pair<T>{
+      thrust::min(x.min_val, y.min_val), true, thrust::max(x.max_val, y.max_val), true};
+  }
+};
+
+/**
+ * @brief functor that calls thrust::transform_reduce to produce a std::pair
+ * of scalars that represent the minimum and maximum values of the input data
+ * respectively. Note that dictionaries and non-relationally comparable objects
+ * are not supported.
+ *
+ */
+struct minmax_functor {
+  template <typename T>
+  // unable to support fixed point due to DeviceMin/DeviceMax not supporting fixed point
+  std::enable_if_t<cudf::is_relationally_comparable<T, T>() and
+                     not std::is_same<T, dictionary32>::value and not cudf::is_fixed_point<T>(),
+                   std::pair<std::unique_ptr<scalar>, std::unique_ptr<scalar>>>
+  operator()(const cudf::column_view &col, rmm::mr::device_memory_resource *mr, cudaStream_t stream)
+  {
+    auto device_col = column_device_view::create(col, stream);
+
+    // compute minimum and maximum values
+    minmax_pair<T> result;
+    if (col.nullable()) {
+      result = thrust::transform_reduce(
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(col.size()),
+        [d_col = *device_col] __device__(size_type index) -> minmax_pair<T> {
+          return minmax_pair<T>(d_col.element<T>(index), d_col.is_valid(index));
+        },
+        minmax_pair<T>{},
+        minmax_with_null_binary_op<T>{});
+    } else {
+      result = thrust::transform_reduce(
```

Comment:
Involves Device->Host (reduction result), Host->Device (scalar creation) transfer.
Is there any way to avoid this?

Reply:
It makes me sad, but I don't know a way around it. Do you have any suggestions?

Reply:
Tried `cub::Reduce`. Benchmarks are compared below.
Time & CPU = (Old-New)/Old
Other columns unit is ns.

23% to 9% reduction for <10M elements.
```
./googlebenchmark/googlebenchmark/tools/compare.py benchmarks ./thrust_reduce.json ./cub_reduce.json
Comparing ./thrust_reduce.json to ./cub_reduce.json
Benchmark                                                             Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------
Reduction/bool_minmax/10000/manual_time                            -0.3858         -0.2313         28474         17489         48015         36910
Reduction/bool_minmax/100000/manual_time                           -0.3892         -0.2333         28788         17583         48098         36876
Reduction/bool_minmax/1000000/manual_time                          -0.3398         -0.2124         32998         21785         51367         40459
Reduction/bool_minmax/10000000/manual_time                         -0.1930         -0.1480         55564         44838         71455         60881
Reduction/bool_minmax/100000000/manual_time                        -0.0502         -0.0464        207780        197345        223835        213453
Reduction/int8_t_minmax/10000/manual_time                          -0.3885         -0.2328         28615         17498         48186         36971
Reduction/int8_t_minmax/100000/manual_time                         -0.3872         -0.2333         28680         17576         47936         36752
Reduction/int8_t_minmax/1000000/manual_time                        -0.3366         -0.2123         32925         21842         51286         40398
Reduction/int8_t_minmax/10000000/manual_time                       -0.1901         -0.1468         55543         44981         71419         60932
Reduction/int8_t_minmax/100000000/manual_time                      -0.0489         -0.0455        207621        197470        223687        213510
Reduction/int32_t_minmax/10000/manual_time                         -0.3911         -0.2308         28238         17195         47698         36689
Reduction/int32_t_minmax/100000/manual_time                        -0.3683         -0.2199         29417         18584         48468         37810
Reduction/int32_t_minmax/1000000/manual_time                       -0.2847         -0.1976         38872         27805         56753         45541
Reduction/int32_t_minmax/10000000/manual_time                      -0.1112         -0.0932        101277         90015        116946        106043
Reduction/int32_t_minmax/100000000/manual_time                     -0.0178         -0.0176        645487        634003        661739        650120
Reduction/timestamp_ms_minmax/10000/manual_time                    -0.4036         -0.2418         28629         17074         48102         36469
Reduction/timestamp_ms_minmax/100000/manual_time                   -0.3676         -0.2306         31785         20099         50524         38872
Reduction/timestamp_ms_minmax/1000000/manual_time                  -0.2192         -0.1673         50422         39369         67050         55833
Reduction/timestamp_ms_minmax/10000000/manual_time                 -0.0735         -0.0657        164089        152028        179949        168119
Reduction/timestamp_ms_minmax/100000000/manual_time                -0.0107         -0.0108       1252399       1238967       1268594       1254885
Reduction/float_minmax/10000/manual_time                           -0.3886         -0.2297         28124         17194         47617         36677
Reduction/float_minmax/100000/manual_time                          -0.3551         -0.2090         29193         18827         48242         38158
Reduction/float_minmax/1000000/manual_time                         -0.2793         -0.1952         38555         27788         56440         45422
Reduction/float_minmax/10000000/manual_time                        -0.1101         -0.0925        101187         90046        116882        106067
Reduction/float_minmax/100000000/manual_time                       -0.0180         -0.0180        645615        634003        661984        650101
```

Reply:
https://github.com/karthikeyann/cudf/commit/36d8daa0488eb7caa01dcf3959e3af81695edb9c This commit contains benchmark and `cub::Reduce` usage. 
(Still there is one HtoD transfer. If only, there is a a way to avoid that too without giving up `stream`).
If this looks fine, Shall I contribute this change to this PR?

Reply:
Is the old time in the benchmark the thrust version of this code? If so, I don't see why we wouldn't change it to use cub since it is certainly faster.

Reply:
Old is `thrust::reduce`. New is `cub::DeviceReduce::Reduce`